### PR TITLE
fix(DiffFlameGraph): Fix the "Explain Flame Graph" (AI) feature

### PR DESCRIPTION
**Pull request by grafakus**
### ✨ Description

**Related issue(s):** is caused by https://github.com/grafana/explore-profiles/pull/119

This PR aims at fixing the "Explain Flame Graph" feature on the comparison view. Namely, before this PR:
- the `isDiff` prop [was not passed](https://github.com/grafana/explore-profiles/pull/119/files#diff-b648e06443c5667d8938e238e6b0a8b7579e793379e4fdd1cfc0c0191a7984e6R182) to the `AiPanel` component, leading to an incorrect AI analysis based solely on a single profile
- obtaining the queries required to fetch the profiles in DOT format relied on `parseLeftRightUrlSearchParams` which worked only for the legacy pages

### 📖 Summary of the changes

To fix the issues, this PR introduces the new `SceneAiPanel` (based on the non-Scenes `AiPanel`), used both in `SceneFlameGraph` and `SceneDiffFlameGraph`.

See the diff tab for specific comments.

### 🧪 How to test?

Manually after checking this PR branch in local:
- The analysis for a single flamegraph should work as before
- The analysis for a diff flamegraph should work

In both case, the API requests made by the browser should have the correct query(-ies) and the correct time range(s). In the case of a single flame graph, the time range must be the last one used to fetch the main timeseries. And in the case of a diff flame graph, they must be equal to the ranges selected by the user on the baseline & comparison timeseries.

### CODE REVIEW

**Comment by grafakus** [comment]
As a note for the next weeks: we need to start adding end-to-end tests to capture these kind of regressions earlier.

File: src/pages/ProfilesExplorerView/components/SceneAiPanel/components/AiButton/AIButton.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  if (error) {
+    icon = 'shield-exclamation';
+    title = 'Grafana LLM plugin missing or not configured!';
+  } else if (isFetching) {
+    icon = 'fa fa-spinner';
+    title = 'Checking the status of the Grafana LLM plugin...';
+  }
+
+  return (
+    <Button
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneAiPanel/components/AiButton/AIButton.tsx:32]: I took the opportunity to remove the "new" badge:

image

The feature has already a few months now.

File: src/pages/ProfilesExplorerView/components/SceneAiPanel/components/AiButton/AIButton.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import { css } from '@emotion/css';
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneAiPanel/components/AiButton/AIButton.tsx:1]: This component, as well as other files have been copied from the "shared" folder and customized for the Scenes app. Once we remove the code for the legacy Comparison pages (which is still present), we'll clean up everything.

File: src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/components/SceneDiffFlameGraph/infrastructure/useFetchDiffProfile.ts [hunk 3]
```diff
@@ -33,10 +33,10 @@ export function useFetchDiffProfile({
     // for UX: keep previous data while fetching -> profile does not re-render with empty panels when refreshing
     placeholderData: (previousData) => previousData,
     enabled,
-    // we use "raw" to cache relative time ranges between renders, so that only refetch() will trigger a new query
     // eslint-disable-next-line @tanstack/query/exhaustive-deps
     queryKey: [
       'diff-profile',
+      dataSourceUid,
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/components/SceneDiffFlameGraph/infrastructure/useFetchDiffProfile.ts:39]: I realised I forgot to add it to the key in the previous PR 🤦🏾‍♂️

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneFlameGraph.tsx [hunk 4]
```diff
@@ -55,7 +60,12 @@ export class SceneFlameGraph extends SceneObjectBase<SceneFlameGraphState> {
 
       dataSubscription = newState.$data?.subscribeToState((newDataState) => {
         if (newDataState.data?.state === LoadingState.Done) {
-          timelineAndProfileApiClient.setLastTimeRange(newDataState.data.timeRange);
+          const lastTimeRange = newDataState.data.timeRange;
+
+          // For the "Function Details" feature only
+          timelineAndProfileApiClient.setLastTimeRange(lastTimeRange);
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneFlameGraph.tsx:66]: We should also migrate the GitHub integration to Scenes at some point in the future.

File: src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/components/SceneDiffFlameGraph/SceneDiffFlameGraph.tsx [hunk 5]
```diff
@@ -141,45 +147,56 @@ export class SceneDiffFlameGraph extends SceneObjectBase<SceneDiffFlameGraphStat
...
-              enableFlameGraphDotComExport={settings?.enableFlameGraphDotComExport}
-              collapsedFlamegraphs={settings?.collapsedFlamegraphs}
+              profile={data.profile}
+              enableFlameGraphDotComExport={data.settings?.enableFlameGraphDotComExport}
+              collapsedFlamegraphs={data.settings?.collapsedFlamegraphs}
             />
           )}
         </Panel>
 
-        {sidePanel.isOpen('ai') && <AiPanel className={styles.sidePanel} onClose={sidePanel.close} />}
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/components/SceneDiffFlameGraph/SceneDiffFlameGraph.tsx:182]: The missing `isDiff` prop caused the feature to be broken.

File: src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/components/SceneDiffFlameGraph/SceneDiffFlameGraph.tsx [hunk 6]
```diff
@@ -88,6 +96,13 @@ export class SceneDiffFlameGraph extends SceneObjectBase<SceneDiffFlameGraphStat
         profile: profile as FlamebearerProfile,
         settings,
         fetchSettingsError,
+        ai: {
+          panel: aiPanel,
+          fetchParams: [
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/components/SceneDiffFlameGraph/SceneDiffFlameGraph.tsx:101]: We now pass the correct parameters (query and time ranges).

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneFlameGraph.tsx [hunk 7]
```diff
@@ -94,18 +112,20 @@ export class SceneFlameGraph extends SceneObjectBase<SceneFlameGraphState> {
...
-        title: `${serviceName} flame graph (${profileMetricType})`,
+        title: this.buildTitle(),
         isLoading: isFetchingProfileData,
         isFetchingProfileData,
         hasProfileData,
         profileData,
         settings,
+        ai: {
+          panel: aiPanel,
+          fetchParams: [{ query, timeRange: lastTimeRange }],
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreServiceFlameGraph/SceneFlameGraph.tsx:127]: It's now easier to understand where the last time range is used and what for (vs calling `  timelineAndProfileApiClient.setLastTimeRange(lastTimeRange)` for the GitHub integration feature).

File: src/pages/ProfilesExplorerView/domain/variables/FiltersVariable/FiltersVariable.tsx [hunk 8]
```diff
@@ -47,18 +46,11 @@ export class FiltersVariable extends AdHocFiltersVariable {
 
   static Component = ({ model }: SceneComponentProps<AdHocFiltersVariable & { onChangeQuery?: any }>) => {
     const styles = useStyles2(getStyles);
+
     const { key } = model.useState();
-    const [, setQuery] = useQueryFromUrl();
 
     const query = useBuildPyroscopeQuery(model, key as string);
 
-    useEffect(() => {
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/domain/variables/FiltersVariable/FiltersVariable.tsx:55]: By removing this code and the one below in `useToggleSidePanel` we are getting rid of the `query` URL search parameters.

**Review by bryanhuhta — APPROVED** [review]
Tested it a variety of ways locally and everything seems to work 👍

File: src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/SceneExploreDiffFlameGraph.tsx [hunk 9]
```diff
@@ -43,6 +44,11 @@ export class SceneExploreDiffFlameGraph extends SceneObjectBase<SceneExploreDiff
   }
 
   onActivate() {
+    // hack to force UrlSyncManager to handle a new location
+    // this will sync the state from the URL by calling updateFromUrl() on all the time ranges (`SceneTimeRange` and our custom `SceneTimeRangeWithAnnotations`) that are defined on `SceneComparePanel`
+    // if not, landing on this view will result in empty URL search parameters (to/from and diffTo/diffFrom) which will make shareable links useless
+    locationService.partial({}, true); // replace to avoid creating history items
```
Comment (grafakus) [inline src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/SceneExploreDiffFlameGraph.tsx:50]: Crazy side effect of removing https://github.com/grafana/explore-profiles/pull/129/files#diff-5db2774cbc84b771946755dd3cc399bf32509129f5bf13d2177950b07f3d6119L58, the URL search parameters for all the time ranges were wiped. 

Does anybody know a better solution (idiomatic to Scenes)?

**Comment by ifrost** [comment]
Played with a bit and it looks good, I'll have a look at the code in a bit. A few comments:

---

**suggestion**: Align max-nodes used for analysis with user settings

We use hardcoded [max-nodes=100](https://github.com/grafana/explore-profiles/blob/1163777b8367dfa511758699eac7af2683da8329/src/shared/components/AiPanel/infrastructure/useFetchDotProfiles.ts#L7) for the analysis. It may yield more (or less) detailed analysis than what's visible in the graph based on user settings. Would it make sense to use user settings and cap it at 100 if uses uses more + some info about how many nodes are being processed?

---

**suggestion**: Refresh or close analysis panel when query changes

It seems like we don't close or refresh the analysis when the profile type changes (only when service changes) which may lead to stale results being shown. Would it make sense to hide the analysis or refresh it automatically when query changes (profile type or filters)?

File: src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/SceneExploreDiffFlameGraph.tsx (hunk 9 above)
Comment (ifrost) [reply src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/SceneExploreDiffFlameGraph.tsx:50]: Is it needed because time range is added dynamically after the URL is synced? Maybe we could use UrlSyncManager.handleNewLocation but we need to upgrade scenes to latest version (which we should do anyway but not part of this PR)

**Review by ifrost — APPROVED** [review]
Nothing blocking from my end but I think we should create follow-ups on code duplication, scene updates and todos

File: src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/SceneExploreDiffFlameGraph.tsx (hunk 9 above)
Comment (grafakus) [reply src/pages/ProfilesExplorerView/components/SceneExploreDiffFlameGraph/SceneExploreDiffFlameGraph.tsx:50]: > Is it needed because time range is added dynamically after the URL is synced?

Yes

**Comment by grafakus** [comment]
Thank you for the suggestions!

> **suggestion**: Align max-nodes used for analysis with user settings
> 
> We use hardcoded [max-nodes=100](https://github.com/grafana/explore-profiles/blob/1163777b8367dfa511758699eac7af2683da8329/src/shared/components/AiPanel/infrastructure/useFetchDotProfiles.ts#L7) for the analysis. It may yield more (or less) detailed analysis than what's visible in the graph based on user settings. Would it make sense to use user settings and cap it at 100 if uses uses more + some info about how many nodes are being processed?

I believe we capped it to 100:
- to limit the number of OpenAI tokens
- because it didn't provide better results with higher values

@aleks-p may is it correct?

> **suggestion**: Refresh or close analysis panel when query changes
> 
> It seems like we don't close or refresh the analysis when the profile type changes (only when service changes) which may lead to stale results being shown. Would it make sense to hide the analysis or refresh it automatically when query changes (profile type or filters)?

We do: https://github.com/grafana/explore-profiles/pull/129/files#diff-950b99b789b4fc4787eb1cf83aa00347e27cd96fbb1dcd6c9ef3bd839a45ba3bR147

File: src/pages/ProfilesExplorerView/components/SceneAiPanel/infrastructure/ProfileApiClient.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import { TimeRange } from '@grafana/data';
+
+import { DataSourceProxyClient } from '../../../infrastructure/series/http/DataSourceProxyClient';
+
+// dot format returns string (TODO: json format later)
```
[the remark this answers is not in the document]
Comment (grafakus) [reply src/pages/ProfilesExplorerView/components/SceneAiPanel/infrastructure/ProfileApiClient.ts:5]: It all depends.

For instance, this one has been tackled in [this PR](https://github.com/grafana/explore-profiles/pull/132/files#diff-f75b69b1b716ea9aaa14b38dadc1a2fc28fb48532caa3381adf764bacdb0736cL11).

But generally speaking, because we gave more importance to speed and flexibility, we were not strict. So far, I've used private reminders to do some code maintenance whenever I could. I now believe we should have some kind of process. Let's talk!

File: src/pages/ProfilesExplorerView/components/SceneAiPanel/domain/useOpenAiChatCompletions.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
+import { llms } from '@grafana/experimental';
```
[the remark this answers is not in the document]
Comment (grafakus) [reply src/pages/ProfilesExplorerView/components/SceneAiPanel/domain/useOpenAiChatCompletions.ts:1]: [AiPanel/domain/useOpenAiChatCompletions.ts](https://github.com/grafana/explore-profiles/blob/1c2095042a4117c58f8fd9e3301797dffe006737/src/shared/components/AiPanel/domain/useOpenAiChatCompletions.ts#L2) is obsolete, so I'll remove it ASAP

File: src/pages/ProfilesExplorerView/components/SceneAiPanel/domain/useOpenAiChatCompletions.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+import { useCallback, useEffect, useState } from 'react';
+import { Subscription } from 'rxjs';
+
+import { buildPrompts, model } from './buildLlmPrompts';
+
+// taken from "@grafana/experimental"
```
[the remark this answers is not in the document]
Comment (grafakus) [reply src/pages/ProfilesExplorerView/components/SceneAiPanel/domain/useOpenAiChatCompletions.ts:7]: This type is not exported.

**Comment by aleks-p** [comment]
> I believe we capped it to 100:
> - to limit the number of OpenAI tokens
> - because it didn't provide better results with higher values
> 
> @aleks-p may is it correct?

Yes, though I think the question is more about starting with the user's max nodes and capping that at 100. This would have an impact if the user is setting max nodes to less than 100 which is fairly unlikely but possible. In this case the AI can "talk" about nodes that the user can't see. 

The issue is a bit more subtle though. The backend determines 2 maxNodes values when making a request for a DOT profile. One is used to retrieve a flame graph and a second one to convert the flame graph to a text (DOT) report (see [this](https://github.com/grafana/pyroscope/blob/ad91b32776e236d383658271a72f44d64e4905ab/pkg/querier/http.go#L153-L162
)).

TLDR: doing something like `min(maxNodes, 100)` could make sense for cases where users configure low values (which I don't think they do). The mapping between flame graph and DOT report nodes is not 1:1 though so there could always be cases where the AI reports things that are not visible to the user.

**Comment by grafakus** [comment]
> > I believe we capped it to 100:
> > 
> > * to limit the number of OpenAI tokens
> > * because it didn't provide better results with higher values
> > 
> > @aleks-p may is it correct?
> 
> Yes, though I think the question is more about starting with the user's max nodes and capping that at 100. This would have an impact if the user is setting max nodes to less than 100 which is fairly unlikely but possible. In this case the AI can "talk" about nodes that the user can't see.
> 
> The issue is a bit more subtle though. The backend determines 2 maxNodes values when making a request for a DOT profile. One is used to retrieve an intermediate flame graph and a second one to convert that flame graph to a text (DOT) report (see [this](https://github.com/grafana/pyroscope/blob/ad91b32776e236d383658271a72f44d64e4905ab/pkg/querier/http.go#L153-L162)).
> 
> TLDR: doing something like `min(maxNodes, 100)` could make sense for cases where users configure low values (which I don't think they do). The mapping between flame graph and DOT report nodes is not 1:1 though so there could always be cases where the AI reports things that are not visible to the user.

Thank you for the clarification!
